### ISSUE TITLE
CustomResource: no metrics if no resource exists for CRD when ksm starts

### ISSUE THREAD

**Issue by chrischdi** [kubernetes/kube-state-metrics#2142]
**What happened**:

There are two issues when no resources exist for a CRD at the time when kube-state-metrics starts:

1. If a resource for a custom resource gets created **after** kube-state-metrics starts:
	- no `HELP` or `TYPE` will get exported
2. In a special case, there will also no metrics get exported for it.

**What you expected to happen**:

* Not requiring a restart to expose the metrics

**How to reproduce it (as minimally and precisely as possible)**:

* Applied a CRD
* Started KSM with configuration for the CRD
* Sleep for 15s
* Created a CR of the above CRD type
* The metrics don't get exposed
* Restart KSM
* Metrics will get exposed

Used Configuration:

```yaml
# This file was auto-generated via: make generate-metrics-config
kind: CustomResourceStateMetrics
spec:
  resources:
  - groupVersionKind:
      group: cluster.x-k8s.io
      kind: Cluster
      version: v1beta1
    labelsFromPath:
      name:
      - metadata
      - name
      namespace:
      - metadata
      - namespace
      uid:
      - metadata
      - uid
    metricNamePrefix: capi_cluster
    metrics:
    - name: created
      help: Unix creation timestamp.
      each:
        gauge:
          path:
          - metadata
          - creationTimestamp
        type: Gauge
    - name: status_condition
      help: The condition of a cluster.
      each:
        stateSet:
          labelName: status
          labelsFromPath:
            type:
            - type
          list:
          - 'True'
          - 'False'
          - Unknown
          path:
          - status
          - conditions
          valueFrom:
          - status
        type: StateSet
```

Note: in this case: after creating the resource:

* The metric `capi_cluster_created` will get exposed, but no `HELP` or `TYPE`.
* The metric for `capi_cluster_status_condition` will not get exposed

**Anything else we need to know?**:

I think the issue got introduced in #1851

If the method `hasResources` which is called here:

https://github.com/kubernetes/kube-state-metrics/blob/cc06755239faf600cc3468737a321204a1432fcf/internal/store/builder.go#L548

get's changed to always return true instead of actually tring to list CRs:

https://github.com/kubernetes/kube-state-metrics/blob/cc06755239faf600cc3468737a321204a1432fcf/internal/store/builder.go#L627

it meets the expectation and metrics get exposed as soon as the first CR for the CRD gets created.

**Environment**:
- kube-state-metrics version: v2.9.2
- Kubernetes version (use `kubectl version`): v1.27.3
- Cloud provider or hardware configuration: kind
- Other info:

**Comment by chrischdi** [comment]
I think the overall issue is in 

https://github.com/kubernetes/kube-state-metrics/blob/bb6e9f42f8bac32ed6e50b6932cb2ab7fc9307ef/pkg/metrics_store/metrics_writer.go#L49

If the first store has no header set: it will add an empty header for the first store

https://github.com/kubernetes/kube-state-metrics/blob/bb6e9f42f8bac32ed6e50b6932cb2ab7fc9307ef/pkg/metrics_store/metrics_writer.go#L63-L65

Afterwards we loop over the headers, and use the index there to write the metrics

https://github.com/kubernetes/kube-state-metrics/blob/bb6e9f42f8bac32ed6e50b6932cb2ab7fc9307ef/pkg/metrics_store/metrics_writer.go#L66-L85

In this case, if we have more than one `metricFamilies` in a single `store` we will only expose the first metric in there.

**Comment by chihshenghuang** [comment]
i have the same issue and provide the log at #2141

**Comment by chrischdi** [comment]
> i have the same issue and provide the log at #2141

Jep, I've seen that issue and also tried the linked PR. Though I was not sure if that describes the same issue.

**Comment by buger** [comment]
> Be aware, geneated by AI architect, may contain some faulty statement

You're right, this is an issue in kube-state-metrics where metrics for custom resources will not be exposed if the custom resource does not exist when kube-state-metrics starts up initially.

The root cause is that kube-state-metrics pre-generates the metric families and headers for custom resources on startup based on the configuration, before any actual resources exist. So if no resources exist, it will not generate any metrics and just have empty metric families.

Then when a custom resource is created after startup, the metrics for that resource will not get picked up properly due to the logic you pointed out:

-   The first metric family in a store having a nil header is replaced with an empty header string.
-   When writing out metrics, it iterates over the headers first and only writes out metrics for existing families.

So any new families added after startup are skipped.

A couple ways this could be fixed:

1.  Don't pre-generate metric families on startup, and instead generate them dynamically as resources are listed and watched. This would require some refactoring to delay family generation until the first List() call.
2.  When a custom resource is created, re-detect headers for any metric stores that had empty headers before writing metrics. This would handle new families being added after startup.
3.  Always write out all metrics, ignoring header iteration order. This assumes stores and families won't get out of sync.

Option 2 might be the least disruptive change, by handling the empty header case better in WriteAll(). Something like:

go

Copy code

`for  _, s :=  range m.stores {   if s.headers ==  nil  &&  len(s.metrics)  >  0  {   // Regenerate headers if they were initially empty  s.headers = generator.ExtractMetricFamilyHeaders(s.generateMetricFamilies)   }
  for i, h :=  range s.headers {   // Write headers   }
  for  _, metricFamilies :=  range s.metrics {   // Write all metrics
  }  }`

This would make sure new families have headers generated correctly before writing metrics.

Here are a few options to overcome the issue of metrics not being exposed for custom resources created after kube-state-metrics has started:

1.  Restart kube-state-metrics whenever you create new custom resources, so it will pick up the new families. Not ideal, but will work.
2.  Modify kube-state-metrics as described above to regenerate headers if empty before writing metrics. This would require building a custom kube-state-metrics image.
3.  Use the `--custom-resource-config` flag to provide the CRD configuration dynamically at startup instead of baked in the image. Then restart kube-state-metrics after creating new CRDs.
4.  Switch to using the Kubernetes Metrics API and scraping the `/apis/custom.metrics.k8s.io/v1beta1` endpoint. This exposes metrics for all custom resources without needing to configure kube-state-metrics. But you lose kube-state-metrics formatting.
5.  File an issue on kube-state-metrics GitHub repo requesting a fix. Could try submitting a PR with the fix too.
6.  Open an issue on Prometheus GitHub repo requesting native support for custom resource metrics. Would solve this issue.
7.  Don't use kube-state-metrics for dynamic custom resources - use built-in Prometheus Kubernetes SD and relabeling to target CRDs.

So in summary, the options are:

-   Restart kube-state-metrics
-   Modify kube-state-metrics yourself
-   Use Metrics API directly
-   Upstream fixes in kube-state-metrics or Prometheus
-   Don't use kube-state-metrics for dynamic CRs

Let me know if this analysis makes sense! I agree it's an issue that should be fixed to properly support dynamic custom resources.

**Comment by chrischdi** [comment]
> > Be aware, geneated by AI architect, may contain some faulty statements

Don't know if I should consider this as spam and report. It does not really provide new and/or useful suggestions which are not already there in the above.

**Comment by buger** [comment]
I woud argue that it does not contain any new info. At least for me, who was not deep into this project before.
Keep in mind that before doing this suggestions, it analysed all the source code, etc. It is not just GPT bot which rewrites original input. 

I would ask not to report, as it just experiement, and I picked a few OSS project to trial my project on real-world problems. And in each project I took up to 5 tickets and PRs. So I'm not going to continue posting this more, unless repository owner will want me to continue.

**Comment by mrueg** [comment]
If you run any further "experiements", please reach out to the repository owners first and ask them for their okay. Spamming a project (even if it's only 5 tickets) without any sort of opt-in is a pretty rude and non-friendly behavior.
This is not helpful to resolve any issue, potentially confusing users and not desired by the maintainers of this project as it wastes time from everyone reading those generated messages.

**To be clear, we do not want you to continue.**

**Comment by ZhangsongLee** [comment]
If we apply a CRD after starting ksm and then create a CR later, The metrics don't get exposed. 
We expect to expose these metrics without restarting ksm。
